### DISCUSSION TITLE
Proposal for supporting type state of query compiling (using Contains)

### DISCUSSION THREAD

**Discussion by ManiMozaffar** [sqlalchemy/sqlalchemy#11087] [Ideas]
Imagine this function:

```python
AnyTuple = TypeVar("AnyTuple", bound=tuple)
AnySelect = sa.Select[AnyTuple]

def query_existence(query_smnt: AnySelect, username: str, email: str) -> AnySelect:
    return query_smnt.where(
        or_(
            sa.func.lower(User.email) == sa.func.lower(email),
            sa.func.lower(User.username) == sa.func.lower(username),
        )
    )
```

This function relies on the fact that User should exists in the select statement.
but it's not type stated, so that's the problem. the caller should always make sure that join is performed, before reaching this function.
but what if..

```python
ContainsUser = TypeVar("AnyTuple", bound=tuple[Contains[User]])
UserSelect = sa.Select[ContainsUser]

def query_existence(query_smnt: UserSelect, username: str, email: str) -> UserSelect:
    return query_smnt.where(
        or_(
            sa.func.lower(User.email) == sa.func.lower(email),
            sa.func.lower(User.username) == sa.func.lower(username),
        )
    )
```

wouldn't it be awesome? Ofc this isn't supported in Python, that's why I'm creating issue, since you guys can come up with a pep and actual proposal. But it's huge win in python type hints in my book.

**Comment by ManiMozaffar** [comment]
note:
tuple[int, ...]

doesn't work in this case. because you may have more in tuple, than just int. Now if you do union, you can't be sure if your tuple is containing your required type. it just checks for all as that what union means.

**Comment by zzzeek** [comment]
>  since you guys can come up with a pep and actual proposal. 

what??!  This is not something that can be done in typing and even if it could, it wouldn't even be correct.  You can most certainly add WHERE criteria that refers to an entity without it being in a JOIN, that's called an "implicit join".     beyond all that , proposing PEPs for even very simple things is an enormous job and we dont really get into that here.

**Comment by ManiMozaffar** [comment]
> what??! This is not something that can be done in typing and even if it could, it wouldn't even be correct. You can most certainly add WHERE criteria that refers to an entity without it being in a JOIN, that's called an "implicit join". beyond all that , proposing PEPs for even very simple things is an enormous job and we dont really get into that here.

yeah I know it's possible via implicit join, tbh I'm not fan of that, rather being explicit. 
I see, I thought it might be possible, it'd be very great anyway for use case of sqla. Thx for responding :)
